### PR TITLE
Add memory arbitration strategy

### PR DESCRIPTION
**Pull request by oerling**
Addd memory arbitration strategy

Adds a process-wide MemoryManagerStrategy instance and an abstract
base class MemoryConsumer. The strategy keeps a list of consumers.
the consumers each have a top level MemoryUsageTracker that gives
their current usage and limit. The consumers have a method for that
returns an estimate of the memory they can free when requested. The
reclaim() method attempts to free a requested amount of memory. The
actually freed memory is reflected in the tracker.

Adds a TaskMemoryStrategy that arbitrates memory asks between
Tasks. The strategy is invoked via the growCallback of the Task top
level MemoryUsageTracker. The top tracker sees the caps of the Task
level trackers as allocations. It is not a direct parent of these
since its allocated bytes reflects the sum of the caps as opposed to
the sum of the allocations.

TaskMemoryStrategy::reclaim first checks if
the process-wide top tracker has capacity and if so, increases the
limit of the requester by the requested amount. If not enough capacity
is available, this does a dirty read of all running Task's reclaimable
memory and sortrs the potential memory donors largest first. This then
pauses the candidates one after the other and forces enough of them to
release memory to meet the demand. The paused tasks are then resumed.

The requested part of the memory ask is used to increase the limit of
the requester and anything left over is added as available in the top
tracker.  Adds a test with a custom operator that keeps growing a
reservation and can be made to decrese the reservation by contending
users.

In the case of a single Task, when any Driver causes the Task total to
go above Task the cap, the strategy is invoked and since there is only one
Task, the Task is stopped and its operators are made to shrink.

Note that in the test the stop and shrink takes place at the initial
reservation of the MemoryComsumer operator. This is a legitimate place
for calling spill on the requesting operator itself. Once the reservation is obtained, we expect the operator to complete processing its batch before checking for pausing. Whenever an operator goes off-thread as a response to a pause request, it must be in a state where spill() can be safely called on it.

However, if it were to exceed its reservation, then it would invoke
the memory arbitration. This in turn could call spill on the calling
operator, which typically would have bad consequences. For example, a
group by would get spilled in the middle of an update, which could
corrupt its state.  Therefore a spill-protect section must be added as
a follow-up.

### CODE REVIEW

**Review by mbasmanova — COMMENTED** [review]
@oerling Overall looks good. Some questions and comments below. Test coverage appears to be thin. Would you check that all functionality is covered? Does this work with grouped execution?

File: velox/exec/Task.cpp [hunk 1]
```diff
@@ -144,9 +144,24 @@ Task::Task(
...
-      bufferManager_(PartitionedOutputBufferManager::getInstance()) {}
+      bufferManager_(PartitionedOutputBufferManager::getInstance()) {
+  auto strategy = memory::MemoryManagerStrategy::instance();
+  if (strategy->canResize()) {
+    auto tracker = memory::MemoryUsageTracker::create(
+        memory::MemoryUsageConfigBuilder().maxTotalMemory(0).build());
+    pool_->setMemoryUsageTracker(tracker);
+    tracker->setGrowCallback([&](memory::MemoryUsageTracker::UsageType type,
+                                 int64_t size,
+                                 memory::MemoryUsageTracker& limit) {
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:155]:
```
s/limit/_tracker/?
```

File: velox/exec/Task.cpp [hunk 2]
```diff
@@ -144,9 +144,24 @@ Task::Task(
       consumerSupplier_(std::move(consumerSupplier)),
       onError_(onError),
       pool_(queryCtx_->pool()->addScopedChild("task_root")),
-      bufferManager_(PartitionedOutputBufferManager::getInstance()) {}
+      bufferManager_(PartitionedOutputBufferManager::getInstance()) {
+  auto strategy = memory::MemoryManagerStrategy::instance();
+  if (strategy->canResize()) {
+    auto tracker = memory::MemoryUsageTracker::create(
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:150]: pool_ should have already been assigned tracker. Can we just set grow callback to it?

File: velox/common/memory/MemoryManagerStrategy.h [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+class MemoryManagerStrategyBase : public MemoryManagerStrategy {
+ public:
+  MemoryManagerStrategyBase() = default;
+
+  bool canResize() const override {
+    return false;
+  }
+
+  void registerConsumer(
+      MemoryConsumer* consumer,
```
Comment (xiaoxmeng) [inline velox/common/memory/MemoryManagerStrategy.h:99]: Can we use task id here as key to consumer?

File: velox/exec/Task.cpp [hunk 4]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
   return it->second.queues;
 }
 
+Driver* FOLLY_NULLABLE Task::thisDriver() const {
+  auto thisThread = std::this_thread::get_id();
+  {
+    std::lock_guard<std::mutex> l(mutex_);
+    for (auto& driver : drivers_) {
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:1460]: NYC: in which case a driver slot could be null? Thanks!

File: velox/exec/Task.cpp [hunk 5]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
...
+      if (driver->state().thread == thisThread) {
+        return driver.get();
+      }
+    }
+  }
+  return nullptr;
+}
+
+int64_t Task::recoverableMemory() const {
+  int64_t total = tracker().maxTotalBytes() - tracker().getCurrentUserBytes();
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:1473]: Shall we init total to available reservation bytes here?

File: velox/exec/Task.cpp [hunk 6]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
...
+  ;
+  for (auto driver : drivers_) {
+    if (driver) {
+      total += driver->recoverableMemory();
+    }
+  }
+  return total;
+}
+
+void Task::recover(int64_t size) {
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/Task.cpp:1483]: +1

File: velox/exec/Task.cpp [hunk 7]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
...
+  available = 0;
+  for (auto [raw, weak] : consumers_) {
+    auto ptr = weak.lock();
+    if (ptr) {
+      auto otherTask = dynamic_cast<Task*>(ptr.get());
+      if (otherTask) {
+        auto size = otherTask->tracker().getCurrentTotalBytes();
+        if (size > kInitialSize) {
+          auto recoverable = otherTask->recoverableMemory();
+          if (recoverable >= 0 /*kMinTransfer*/) {
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:1579]:
```
if (recoverable > 0)? If it is zero, then we can revoke any memory from this task?
```

File: velox/exec/Task.cpp [hunk 8]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
...
+    }
+  }
+  std::sort(
+      candidates.begin(),
+      candidates.end(),
+      [&](const ConsumerScore& left, const ConsumerScore& right) {
+        // Most available first.
+        return left.available > right.available;
+      });
+  if (available < minSize) {
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/Task.cpp:1595]: +1

File: velox/exec/Task.cpp [hunk 9]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
...
+      auto& taskTracker = task->tracker();
+      auto preUsage = taskTracker.maxTotalBytes();
+      task->recover(sizeToGo);
+      auto postUsage = taskTracker.getCurrentTotalBytes();
+      auto delta = preUsage - postUsage;
+      if (delta < kMinTransfer) {
+        // Too little reclaimed. No change to limit.
+        continue;
+      }
+      delta = roundDown(delta, kMinTransfer);
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:1621]:
```
nit: how about kMinReclaimBytes/?
```

File: velox/exec/Task.cpp [hunk 10]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
...
+      auto& exec = folly::QueuedImmediateExecutor::instance();
+      std::move(future).via(&exec).wait();
+      if (task->state() != TaskState::kRunning) {
+        continue;
+      }
+      VELOX_CHECK_EQ(0, task->numThreads());
+      paused.push_back(std::static_pointer_cast<Task>(candidate.consumer));
+
+      auto& taskTracker = task->tracker();
+      auto preUsage = taskTracker.maxTotalBytes();
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:1613]:
```
nit: s/preUsage/currentMemLimit/?
```

File: velox/exec/Task.cpp [hunk 11]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
...
+      delta = roundDown(delta, kMinTransfer);
+      VLOG(1) << fmt::format(
+          "{} drops limit by {} to {}",
+          task->taskId(),
+          delta,
+          taskTracker.maxTotalBytes() - delta);
+      setLimit(taskTracker.maxTotalBytes() - delta, taskTracker);
+      recovered += delta;
+
+      if (recovered >= minSize) {
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:1630]: nit: looks like we don't need sizeToGo? minSize - recovered is sizeToGo?

File: velox/exec/Task.cpp [hunk 12]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
...
+      [&](const ConsumerScore& left, const ConsumerScore& right) {
+        // Most available first.
+        return left.available > right.available;
+      });
+  if (available < minSize) {
+    return false;
+  }
+  int64_t sizeToGo = minSize;
+  int64_t recovered = 0;
+  std::vector<std::shared_ptr<Task>> paused;
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:1600]:
```
nit: s/paused/pausedTasks/
```

File: velox/exec/Task.cpp [hunk 13]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
...
+      if (recovered >= minSize) {
+        break;
+      }
+      sizeToGo -= delta;
+    }
+  }
+  bool success = recovered >= minSize;
+  // Updating downward, no throw.
+  if (success) {
+    topTracker->update(minSize - recovered);
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:1639]:
```
nit: s/minSize/numBytesToReclaim or numBytesToRevoke/
```

File: velox/exec/Driver.cpp [hunk 14]
```diff
@@ -613,15 +620,57 @@ SuspendedSection::SuspendedSection(Driver* FOLLY_NONNULL driver)
...
+
+bool Driver::growTaskMemory(
+    memory::MemoryUsageTracker::UsageType type,
+    int64_t size,
+    memory::MemoryUsageTracker& tracker) {
+  bool result;
+  SuspendedSection::suspended(this, [&]() {
+    auto current = tracker.totalReservedBytes();
+    auto limit = tracker.maxTotalBytes();
+    if (current < limit) {
```
Comment (xiaoxmeng) [inline velox/exec/Driver.cpp:646]: Shall we check current + size <= limit?
s/size/growBytes/

File: velox/exec/Operator.h [hunk 15]
```diff
@@ -366,6 +366,26 @@ class Operator {
     return stats_.planNodeId;
   }
 
+  bool reserveAndRun(
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/Operator.h:369]: +1

**Review by xiaoxmeng — COMMENTED** [review]
Looks good overall. Some minors. Haven't looked at the test yet. Thanks!

**Comment by oerling** [comment]
Thank you for the review, addressed the comments. Answer to questions below:

 

 

 

+  int64_t recoverableMemory() const;

s/recoverableMemory/revocableMemory/ is more accurate in this context? Thanks!

 

 

 

A: Considering the meaning of the words, revoke' means to take away a right or privilege. It is on or off. Recover means among other things to re-acquire something lost, e.g. the part of the stolen money was recovered.' So, I'd say recover is better since it implies a scale rather than all or none.

 

 

+  topTracker->update(-tracker().maxTotalBytes());

Can we let the tracker dtor to release the bytes and propagate to its parent?

 

 

A: the top tracker is not the parent of the task tracker.

If the top tracker were the parent, memory recovered in the grow callback would be posted there. This is not wanted because then somebody else could get it from there. Instead, the grow callback gathers memory from tasks and does ont put it back into anything except at the end where it divides it betwen between the requested and top tracker.

 

 

 

 

A:

+  int64_t minSize =

Why we set minSize to kInitialSize if tracker.maxTotalBytes() < kMinTransfer? Thanks!

 

A: A task in this model starts with 0 reservation. Its first action is to call grow callback to get itself something. The initial request is rounded up for the purpose. The thread that makes the Task cannot go via the grow callback to the arbiter because only threads of the Task itself can do that. So getting the initial reservation via the growth route is most consistent and least special cases. The first Driver gets the initial reservation. The next might queue up to do the same but when it is its turn it will see the reservation is already increased.

 

 

pool_ should have already been assigned tracker. Can we just set grow callback to it?

 

 

 

A: We want a tracker that is a child of any other tracker. There is no such guarantee for the pre-existing one, if any.

 

 

+      MemoryConsumer* consumer,

Can we use task id here as key to consumer?

 

 

 

A:The pointer to the Task is guaranteed unique whereas the id is not necessarily so. The pointer is the most compact

 

 

+    for (auto& driver : drivers_) {

NYC: in which case a driver slot could be null? Thanks!

 

 

 

A: During destruction or after the Driver is finished it could be nullptr, so need to check.

 

+  int64_t total = tracker().maxTotalBytes() - tracker().getCurrentUserBytes();

Shall we init total to available reservation bytes here?

 

 

 

A: The Task tracker is the sum of the reservations of child trackers. So the current user bytes is the sum that we want, it does not have a reservation of its own, rather the subsidiary reservations are added up and counted as child allocations. Reservation at the child means a guarantee to  be able to allocate so much, so to the parent it counts as already allocated.

 

 

+          if (recoverable >= 0 /*kMinTransfer*/) {

if (recoverable > 0)? If it is zero, then we can revoke any memory from this task?

 

 

A: If there is less than min transfer to get from a Task, then there is no point trying to get anything.

 

 

+      delta = roundDown(delta, kMinTransfer);

nit: how about kMinReclaimBytes/?

 

A: The constant means the smallest amount by which it makes sense to shrink a Task. The practical size is the full spill of one operator. This is, supposing 1GB tasks, around 50MB, supposing 20 threads per task.  The minimum should be small enough,~10MB.

**Comment by oerling** [comment]
Meng, Masha 

 

I expect to still rearrange the files a bit, like extracting and renaming the arbiter class.  

 

Thanks

Orri

File: velox/exec/Task.cpp [hunk 16]
```diff
@@ -1442,6 +1464,220 @@ Task::getLocalExchangeQueues(
...
+      std::move(future).via(&exec).wait();
+      if (task->state() != TaskState::kRunning) {
+        continue;
+      }
+      VELOX_CHECK_EQ(0, task->numThreads());
+      pausedTasks.push_back(std::static_pointer_cast<Task>(candidate.consumer));
+
+      auto& taskTracker = task->tracker();
+      auto previousBytes = taskTracker.maxTotalBytes();
+      task->recover(bytesForRequester - recovered);
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:1638]: Ok. The algorithm here is to reclaim as much as possible from the first candidate which has most revocable memory? Thanks!

File: velox/exec/Task.cpp [hunk 17]
```diff
@@ -1442,6 +1464,220 @@ Task::getLocalExchangeQueues(
...
+        consumerTask->taskId(),
+        bytesForRequester,
+        tracker.maxTotalBytes() + bytesForRequester);
+    setLimit(tracker.maxTotalBytes() + bytesForRequester, tracker);
+  } else {
+    topTracker->update(-recovered);
+  }
+
+  for (auto& task : pausedTasks) {
+    if (task->state() == TaskState::kRunning) {
```
Comment (xiaoxmeng) [inline velox/exec/Task.cpp:1674]: Is it thread safe? A concurrent thread can terminate the task between the state check and resume? Shall we instead to check for termination in resume method? Not sure if my understanding is correct or not. But we could leave it for followup.

File: velox/exec/Task.cpp [hunk 18]
```diff
@@ -1432,6 +1453,208 @@ Task::getLocalExchangeQueues(
...
+  return total;
+}
+
+void Task::recover(int64_t size) {
+  VELOX_CHECK_EQ(numThreads_, 0, "recover expects paused task");
+  int32_t numDrivers = 0;
+  for (auto& driver : drivers_) {
+    numDrivers += driver != nullptr;
+  }
+  if (!numDrivers) {
```
[the remark this answers is not in the document]
Comment (xiaoxmeng) [reply velox/exec/Task.cpp:1501]: +1. If not possible, shall we add a CHECK here instead?

File: velox/exec/tests/DriverTest.cpp [hunk 19]
```diff
@@ -804,6 +834,203 @@ TEST_F(DriverTest, driverCreationThrow) {
...
+  RowVectorPtr getOutput() override {
+    if (!input_) {
+      return nullptr;
+    }
+    int64_t size = 10 << 20;
+    if (!reserveAndRun(
+            recoverableTracker_,
+            size,
+            [this](int64_t size) {
+              LOG(INFO) << "Spiller called: " << size;
```
Comment (xiaoxmeng) [inline velox/exec/tests/DriverTest.cpp:872]: Remove this info log?

File: velox/exec/Operator.cpp [hunk 20]
```diff
@@ -361,4 +361,47 @@ void OperatorStats::clear() {
   runtimeStats.clear();
 }
 
+namespace {
+bool tryReserveAndRun(
```
Comment (xiaoxmeng) [inline velox/exec/Operator.cpp:365]: Not sure if we need this abstraction. But reserveAndRun can be in a separate PR if we need in production code? We can implement this method in DriverTest for test purpose?

File: velox/common/memory/MemoryManagerStrategy.cpp [hunk 21]
```diff
@@ -0,0 +1,41 @@
...
+#include <folly/executors/thread_factory/InitThreadFactory.h>
+
+namespace facebook::velox::memory {
+
+bool MemoryManagerStrategy::initialized_;
+MemoryManagerStrategy::MemoryManagerStrategyFactory
+    MemoryManagerStrategy::factory_;
+
+// static
+MemoryManagerStrategy* MemoryManagerStrategy::instance() {
```
Comment (xiaoxmeng) [inline velox/common/memory/MemoryManagerStrategy.cpp:30]: We could use system config to configure the strategy to use, and store the strategy object as a static object? Something like:
```
MemoryManagerStrategy* MemoryManagerStrategy::instance() {
  static std::unique_ptr<MemoryManagerStrategy> strategy = createMemoryStrategy(config option);
  return strategy.get();
}

createMemoryStrategy(strategyName) {
   if (factoryMap_.find(strategyName) != factoryMap_.end()) {
      return factoryMap_[strategyName].create();
   } else {
      return createDefault();
   }
}
```

File: velox/exec/tests/DriverTest.cpp [hunk 22]
```diff
@@ -804,6 +834,203 @@ TEST_F(DriverTest, driverCreationThrow) {
...
+            std::dynamic_pointer_cast<const TestingConsumerNode>(node)) {
+      return maxDrivers_;
+    }
+    return std::nullopt;
+  }
+
+ private:
+  uint32_t maxDrivers_;
+  std::atomic<int32_t>& sequence_;
+  DriverTest* testInstance_;
```
Comment (xiaoxmeng) [inline velox/exec/tests/DriverTest.cpp:948]: testInstance_ is not used?

File: velox/exec/tests/DriverTest.cpp [hunk 23]
```diff
@@ -804,6 +834,203 @@ TEST_F(DriverTest, driverCreationThrow) {
...
+    }));
+  }
+  for (int32_t i = 0; i < kNumTasks; ++i) {
+    threads[i].join();
+    EXPECT_EQ(counters[i], kThreadsPerTask * hits);
+    EXPECT_TRUE(stateFutures_.at(i).isReady());
+  }
+  executor->join();
+  // The limits are released into topTracker at destruction.
+  tasks_.clear();
```
Comment (xiaoxmeng) [inline velox/exec/tests/DriverTest.cpp:1030]: Consider to add revocation stats to verify in test in followup.

**Review by xiaoxmeng — APPROVED** [review]
The overall approach looks good to me. Thanks for putting all these together! We might need to improve the implementation a bit in followup per our offline discussion. Please also have @mbasmanova to take a look!

**Comment by xiaoxmeng** [comment]
> available reser

Thanks for the explanation!

File: velox/exec/Operator.cpp (hunk 20 above)
Comment (mbasmanova) [reply velox/exec/Operator.cpp:365]: @oerling @xiaoxmeng has a good point.  Looks like we don't need to add reserveAndRun method to the Operator interface, do we?

File: velox/common/memory/Memory.h [hunk 24]
```diff
@@ -550,6 +551,12 @@ class IMemoryManager {
   virtual bool reserve(int64_t size) = 0;
   // Subtracts from current total and regain memory quota.
   virtual void release(int64_t size) = 0;
+  virtual void registerConsumer(
```
[the remark this answers is not in the document]
Comment (mbasmanova) [reply velox/common/memory/Memory.h:554]: +1 @oerling Please, document these API and add tests.

**Review by mbasmanova — COMMENTED** [review]
@oerling Orri, have you tried running this on a cluster? The PR includes just one test, which feels insufficient for a complex change like this.

> We might need to improve the implementation a bit in followup per our offline discussion. 

@xiaoxmeng Would you summarize this discussion to clarify what enhancements are planned? It would be nice to explain the gaps in current implementation and planned enhancements in the PR description.

**Comment by mbasmanova** [comment]
@oerling The PR description talks about arbitrating among Tasks, but doesn't explain who the same works for a single Task with multiple operators. Sapphire-on-Velox will use only one task, but run it multi-threaded. I assume we'll have multiple threads running same kind of operator. Would you describe how things work in that use cases? Can we have add a test that exercises this code path?

**Comment by oerling** [comment]
Thank you for the review, addressed the comments, details follow where applicable.

 

+  if (!numDrivers) {

When would this happen?

 

 

A: If the Task were shutting down. We could get references from the collection in the arbiter object.

 

 

+    // Per driver recovery target is 16MB + target / number of drivers.

Why 16MB ? What does "per driver recovery target" means? Different drivers may be running different pipelines, difference operators, hence, may have different memory usage.

 

 

 

A: There is a certain minimum below which recovering memory does not make sense. In practice The recovery consists of spilling the whole operator. The question is, how many operators have to be wholly spilled to make the space? Ch. comment.

 

+void Task::recover(int64_t size) {

Would it make sense for this method to return number of recovered bytes?

 

A: The actionable quantity is the level in the Task's tracker. For clarity we should refer to the level in the tracker as opposed to reading it and subtracting from it.

 

 

Shouldn't we use 'bytes' instead of 'minSize' here?

 

 

A: We use the rounded-up quantity minSize.

 

Do we allow the application to decide which operators should spill?

 

 

A: No. You could say that the preference would be order by first, group by next and join last. Join is a bit different because it must spill on all drivers in a coordinated way. This is possible though since at the time the external induced spill happens the task is stopped. If the task itself decides to spill then the deciding driver must stop the other ones. We'll get back to this  with hash join spill.

 

Generally the application could have a way to express preferences on spill choice but we need use cases to  see how this would go.
